cli_tools/google-osconfig-agent/ospatch/watcher.go

```diff
@@ -47,78 +46,100 @@ func updateEtag(resp *http.Response) bool {
 	return etag != oldEtag
 }
 
-func watchMetadata() (string, error) {
+type watchMetadataRet struct {
+	attr *attributesJSON
+	err  error
+}
+
+type attributesJSON struct {
+	PatchNotify string `json:"osconfig-patch-notify"`
+}
+
+func watchMetadata(c chan watchMetadataRet) {
 	client := &http.Client{}
 
 	req, err := http.NewRequest("GET", metadataURL+etag, nil)
 	if err != nil {
-		return "", err
+		c <- watchMetadataRet{
+			attr: nil,
+			err:  err,
+		}
 	}
 	req.Header.Add("Metadata-Flavor", "Google")
 	resp, err := client.Do(req)
 	if err != nil {
-		return "", err
+		c <- watchMetadataRet{
+			attr: nil,
+			err:  err,
+		}
 	}
 
 	updateEtag(resp)
 
 	md, err := ioutil.ReadAll(resp.Body)
 	resp.Body.Close()
 	if err != nil {
-		return "", err
+		c <- watchMetadataRet{
+			attr: nil,
+			err:  err,
+		}
 	}
 
-	return string(md), nil
+	var metadata attributesJSON
+	err = json.Unmarshal(md, &metadata)
+
+	c <- watchMetadataRet{
+		attr: &metadata,
+		err:  err,
+	}
 }
 
-func watcher(ctx context.Context, savedPatchJobName string) {
+func watcher(ctx context.Context, savedPatchJobName string, cancel <-chan struct{}) {
 	currentPatchJobName = savedPatchJobName
 	webError := 0
 	for {
-		_, err := watchMetadata()
-		if err != nil {
-			// Only log the second web error to avoid transient errors and
-			// not to spam the log on network failures.
-			if webError == 1 {
-				if urlErr, ok := err.(*url.Error); ok {
-					if _, ok := urlErr.Err.(*net.DNSError); ok {
-						logger.Errorf("DNS error when requesting metadata, check DNS settings and ensure metadata.internal.google is setup in your hosts file.")
-					}
-					if _, ok := urlErr.Err.(*net.OpError); ok {
-						logger.Errorf("Network error when requesting metadata, make sure your instance has an active network and can reach the metadata server.")
+		c := make(chan watchMetadataRet)
+		go func(c chan watchMetadataRet) {
+			watchMetadata(c)
+		}(c)
+
+		select {
+		case <-cancel:
+			return
+		case ret := <-c:
+			if ret.err != nil {
+				// Only log the second web error to avoid transient errors and
+				// not to spam the log on network failures.
+				if webError == 1 {
+					if urlErr, ok := ret.err.(*url.Error); ok {
+						if _, ok := urlErr.Err.(*net.DNSError); ok {
+							logger.Errorf("DNS error when requesting metadata, check DNS settings and ensure metadata.internal.google is setup in your hosts file.")
+						}
+						if _, ok := urlErr.Err.(*net.OpError); ok {
+							logger.Errorf("Network error when requesting metadata, make sure your instance has an active network and can reach the metadata server.")
+						}
 					}
+					logger.Errorf(ret.err.Error())
```

Comment:
This will lead to double logs. Should there be a `continue` after the previous log statements?